### PR TITLE
Propose Documentation for Software Best Practices, Coding Conventions, and Style

### LINKED ISSUE #335: Propose Documentation for Software Best Practices, Coding Conventions, and Style
#### Problem
CHIP will be composed of much software contributed from participating member organizations. However, for any new CHIP code, there should be documentation for software best practices, coding conventions, and style.

#### Proposed Solution
Submit a strawman documentation proposal for such software best practices, coding conventions, and style.

### PR DESCRIPTION
**Pull request by gerickson**
#### Problem
CHIP will be composed of much software contributed from participating member organizations. However, for any new CHIP code, there should be documentation for software best practices, coding conventions, and style.

#### Summary of Changes
Proposed a minimal framework for a documentation guide, in AsciiDoc format, for software best practices, coding conventions, and style.

This reflects [this conversation snippet on a prior, broader-scope approach to this document](https://github.com/project-chip/connectedhomeip/pull/358#issuecomment-652597203) in that it restricts the initial framework to essentially one simple maxim: "When in Rome, do as the Romans do.". We can add from here as repetition and consensus dictate.

Fixes #335

### CODE REVIEW

**Review by rwalker-apple — APPROVED** [review]
lgtm

**Review by saurabhst — APPROVED** [review]

File: docs/style/coding/CODING_STYLE_GUIDE.adoc [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+link:#id.jzphr1iiku89[Figure 1 below] attempts to illustrate both
+qualitative and quantitative applicability of these guidelines to
+Project Connected Home over IP software.
+
+Generally, product-specific applications have the greatest flexibility
+and latitude in applying these guidelines to their software. Whereas,
+shared infrastructure bears the least flexibility and bears the greatest
+adherence to these guidelines.
+
+image:CODING_STYLE_GUIDE-figure1.png[Figure 1. Graphical summary of the
```
Comment (woody-apple) [inline docs/style/coding/CODING_STYLE_GUIDE.adoc:83]: I'm not sure that we should say that C++ is required, and straight C is not ok? (Maybe I'm reading this diagram incorrectly).

Maybe we can just remove this diagram, and/or make C allowed in all boxes?

**Review by woody-apple — CHANGES_REQUESTED** [review]

**Review by BroderickCarlin — APPROVED** [review]

File: docs/style/coding/CODING_STYLE_GUIDE.adoc (hunk 1 above)
Comment (gerickson) [reply docs/style/coding/CODING_STYLE_GUIDE.adoc:83]: Updated the diagram to show / allow C/C++ in C++ contexts.

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
@jelderton @hawk248 @andy31415 ?

**Review by andy31415 — APPROVED** [review]

**Pull request merged**